pkg/node/node.go

```diff
@@ -424,41 +429,157 @@ func (n *Node) publishSignature(
 			continue
 		}
 
-		logger.Infof("signature submitted for keep [%s]", keepAddress.String())
+		if !(n.waitForSignature(keepAddress, digest) && n.confirmSignature(keepAddress, digest)) {
+			time.Sleep(retryDelay) // TODO: #413 Replace with backoff.
+			continue
+		}
+
 		return nil
 	}
 }
 
-// monitorKeepPublicKeySubmission observes the chain until either the first
-// conflicting public key is published or until keep established public key
-// or until key generation timed out.
-func (n *Node) monitorKeepPublicKeySubmission(
-	abort chan interface{},
+func (n *Node) waitForSignature(
 	keepAddress common.Address,
-) {
-	monitoringCtx, monitoringCancel := context.WithTimeout(
-		context.Background(),
-		monitorKeepPublicKeySubmissionTimeout,
+	digest [32]byte,
+) bool {
+	const waitTimeout = 10 * time.Minute
+	const checkTick = 1 * time.Minute
+
+	ctx, cancelCtx := context.WithTimeout(context.Background(), waitTimeout)
+	defer cancelCtx()
+
+	checkTicker := time.NewTicker(checkTick)
+	defer checkTicker.Stop()
+
+	logger.Infof(
+		"starting waiting for signature for keep [%s]",
+		keepAddress.String(),
 	)
-	defer monitoringCancel()
 
-	publicKeyPublished := make(chan *eth.PublicKeyPublishedEvent)
-	conflictingPublicKey := make(chan *eth.ConflictingPublicKeySubmittedEvent)
+	for {
+		select {
+		case <-checkTicker.C:
+			isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(
```

Comment:
Can we add a comment - same as for public key submission - why do we prefer not to use events in this case?
```
// We check the state periodically instead of relying on
// incoming events. The main motivation is that events could not be
// trusted here because they may come from a forked chain or can
// the same event can be delivered multiple times.
```
?